src/analysis/processing/qgsalgorithmfillnodata.cpp

```diff
@@ -0,0 +1,158 @@
+                      "The algorithm respects the input raster data type (eg. a fill value with commas gets truncated "
+                      "when applied to an integer raster)." );
+}
+
+QgsFillNoDataAlgorithm *QgsFillNoDataAlgorithm::createInstance() const
+{
+  return new QgsFillNoDataAlgorithm();
+}
+
+bool QgsFillNoDataAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( feedback );
+  mInputRaster = parameterAsRasterLayer( parameters, QStringLiteral( "INPUT" ), context );
+  mFillValue = parameterAsDouble( parameters, QStringLiteral( "FILL_VALUE" ), context );
+
+  if ( !mInputRaster )
+    throw QgsProcessingException( invalidRasterError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  mBand = parameterAsInt( parameters, QStringLiteral( "BAND" ), context );
+  if ( mBand < 1 || mBand > mInputRaster->bandCount() )
+    throw QgsProcessingException( QObject::tr( "Invalid band number for BAND (%1): Valid values for input raster are 1 to %2" ).arg( mBand ).arg( mInputRaster->bandCount() ) );
+
+  mInterface.reset( mInputRaster->dataProvider()->clone() );
+  mInputNoDataValue = mInputRaster->dataProvider()->sourceNoDataValue( mBand );
+  mExtent = mInputRaster->extent();
+  mLayerWidth = mInputRaster->width();
+  mLayerHeight = mInputRaster->height();
+  mCrs = mInputRaster->crs();
+  mNbCellsXProvider = mInterface->xSize();
+  mNbCellsYProvider = mInterface->ySize();
+  return true;
+}
+
+QVariantMap QgsFillNoDataAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  //test if input dataset has NoData
+  if ( !mInputRaster->dataProvider()->sourceHasNoDataValue( mBand ) )
+    throw QgsProcessingException( QObject::tr( "Input raster has no NoData values. There exist no NoData cells to fill." ) );
+
+  //prepare output dataset
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+  std::unique_ptr< QgsRasterFileWriter > writer = qgis::make_unique< QgsRasterFileWriter >( outputFile );
+  writer->setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer->setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider > provider( writer->createOneBandRaster( mInterface->dataType( mBand ), mNbCellsXProvider, mNbCellsYProvider, mExtent, mCrs ) );
+  if ( !provider )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( outputFile ) );
+  if ( !provider->isValid() )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output %1: %2" ).arg( outputFile, provider->error().message( QgsErrorMessage::Text ) ) );
+
+  //prepare output provider
+  mDestinationRasterProvider = provider.get();
+  mDestinationRasterProvider->setNoDataValue( 1, -9999 );
```

Comment:
I don't think we want to set this -- the output is supposed to have NO nodata value, and potentially -9999 is an existing, valid value used within the input raster layer.